### PR TITLE
stm32 ringbuffered adc docs

### PR DESCRIPTION
**Pull request by andresv**

### CODE REVIEW

**Comment by sethkrie** [comment]
If `measurements` has length N and DMA buffer has length M:
 - N should be <= M/2  
 - Divisible by sequence length

Everytime read_exact() wakes, there will be at least M/2 new samples (half or full transfer IR fires). We want to grab as many as we can so as to avoid overrun and busy looping.

Say you have M = 120, but set N = 3
> ADC and DMA start
> User calls .read_exact()
> .poll() until half transfer IR (60 new samples)
> .read_exact() returns [u16;3], 57 new samples waiting in DMA buf.
> .read_exact() called again
> .poll() till transfer complete IR (60 new samples
> .read_exact() returns [u16;3], 114 samples waiting
> Next iter = overrun

**Comment by andresv** [comment]
So it does not wake when first batch of channels are sampled and it only wakes for half or full transfer?

**Comment by sethkrie** [comment]
That's correct, since .read_exact() is using the ReadableRingBuffer as the waker, not EOCS.

Here's what the ReadableRingBuffer docs say about the function:

> Async/Wake Behavior: The underlying DMA peripheral only can wake us when its buffer pointer has reached the halfway point, and when it wraps around.

**Comment by andresv** [comment]
Wouldn't it be correct to always use `N == M/2` and not just `N <= M/2`? In my example N = 3 and M = 120 and as you mentioned this actually ends up with an overflow.

**Comment by sethkrie** [comment]
Yep, now that I think about it, you're 100% right.  If ever a buffer  `<M/2` is used, overrun will occur eventually.

That can be enforced using asserts or generic args like you have in your v4 PR. It'd be easy enough because the ringbuf ADC struct owns ReadableRingBuffer, which knows it's own size.

**Comment by sethkrie** [comment]
```
assert_eq!(
    self.ring_buf.capacity() / 2,
    N,
    "Buffer size must be half the size of the ring buffer"
);
```

Not sure if this belongs in a docs PR though...

**Comment by andresv** [comment]
Discussion is all here. Added also assert to this PR and updated docs.

@sethkrie could you check them and if they more or less make sense now.

File: embassy-stm32/src/adc/ringbuffered_v2.rs [hunk 1]
```diff
@@ -341,7 +356,64 @@ impl<'d, T: Instance> RingBufferedAdc<'d, T> {
...
+    ///
+    /// This method fills the provided `measurements` array with ADC readings.
+    /// The length of the `measurements` array should be exactly half of the DMA buffer length.
+    /// Because interrupts are only generated if half or full DMA transfer completes.
+    ///
+    /// Each call to `read_exact` will populate the `measurements` array in the same order as the channels defined with `set_sample_sequence`.
+    /// There will be many sequences worth of measurements in this array because it only returns if at least half of the DMA buffer is filled.
+    /// For example if 3 channels are sampled `measurements` contain: `[sq0 sq1 sq3 sq0 sq1 sq3 sq0 sq1 sq3 sq0 sq1 sq3..]`.
+    ///
+    /// If an error is returned, it indicates a DMA overrun, and the process must be restarted by calling `start` again.
```
[the remark this answers is not in the document]
Comment (andresv) [reply embassy-stm32/src/adc/ringbuffered_v2.rs]: I totally missed that, added a comment about this.

File: embassy-stm32/src/adc/ringbuffered_v2.rs [hunk 2]
```diff
@@ -91,6 +91,15 @@ pub struct RingBufferedAdc<'d, T: Instance> {
 }
 
 impl<'d, T: Instance> Adc<'d, T> {
+    /// Configures the ADC to use a DMA ring buffer for continuous data acquisition.
+    ///
+    /// The `dma_buf` should be large enough to prevent buffer overflow, allowing sufficient time to read out measurements.
```
[the remark this answers is not in the document]
Comment (andresv) [reply embassy-stm32/src/adc/ringbuffered_v2.rs]: Fixed.

File: embassy-stm32/src/adc/ringbuffered_v2.rs [hunk 3]
```diff
@@ -341,7 +357,54 @@ impl<'d, T: Instance> RingBufferedAdc<'d, T> {
         }
     }
 
-    pub async fn read_exact<const N: usize>(&mut self, buf: &mut [u16; N]) -> Result<usize, OverrunError> {
+    /// Reads measurements from the DMA ring buffer.
+    ///
+    /// This method fills the provided `measurements` array with ADC readings. The length of the `measurements` array should match the number of ADC channels.
+    /// For example, if there are 3 ADC channels, then `measurements` should be an array of length 3.
```
[the remark this answers is not in the document]
Comment (andresv) [reply embassy-stm32/src/adc/ringbuffered_v2.rs]: Fixed.

**Comment by andresv** [comment]
@sethkrie fixed your comments. Give thumbs up if it is ok now.

**Review by sethkrie — APPROVED** [review]

**Comment by sethkrie** [comment]
Thanks Andres!

**Comment by andresv** [comment]
@Dirbaio this is done and can be merged.

**Pull request merged**